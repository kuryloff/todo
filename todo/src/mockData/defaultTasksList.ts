import { Task } from "../interfaces/todoList.interface";
//@ts-ignore
import { v4 as uuidv4 } from "uuid";

export const defaultTasksList: Task[] = [
  {
    title: "Tasks creation management",
    id: uuidv4(),
    description: "Add and remove tasks",
    status: 0,
    subtasks: [
      {
        id: uuidv4(),
        title: "Add tasks",
        status: 0,
      },
      {
        id: uuidv4(),
        title: "Remove tasks",
        status: 0,
      },
    ],
  },
  {
    title: "Additional display info",
    id: uuidv4(),
    description: "Display a count of remaining incomplete tasks",
    status: 0,
    subtasks: [
      {
        id: uuidv4(),
        title: "Show counter",
        status: 1,
      },
      {
        id: uuidv4(),
        title: "Apply counter function",
        status: 0,
      },
    ],
  },
  {
    title: "Local data saving",
    id: uuidv4(),
    description: "Data should persisted using some form of browser storage",
    status: 0,
    subtasks: [
      {
        id: uuidv4(),
        title: "Create storage",
        status: 0,
      },
      {
        id: uuidv4(),
        title: "Update storage functionality",
        status: 0,
      },
    ],
  },
  {
    title: "Tasks display",
    id: uuidv4(),
    description: "Should display as a list of Task cards",
    status: 0,
    subtasks: [
      { id: uuidv4(), title: "Create Task Card", status: 0 },
      { id: uuidv4(), title: "Arrange display", status: 0 },
    ],
  },
  {
    title: "testing",
    id: uuidv4(),
    description: "Should have unit tests",
    status: 0,
    subtasks: [],
  },
  {
    title: "Task completion",
    id: uuidv4(),
    description:
      "Should be able to mark a task status as complete by clicking a checkbox",
    status: 0,
    subtasks: [{ id: uuidv4(), title: "Create update functions", status: 0 }],
  },
  {
    title: "Update",
    id: uuidv4(),
    description: "Should be able to update all criteria of a task",
    status: 0,
    subtasks: [
      {
        id: uuidv4(),
        title: "Should show title as crossed out when complete",
        status: 0,
      },
      { id: uuidv4(), title: "Arrange display", status: 0 },
    ],
  },
];
