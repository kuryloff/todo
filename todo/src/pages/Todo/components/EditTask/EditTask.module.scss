@import "src/styles/mixins";
.container{
  width: 40%;
}
.headerTitle{
  text-align: center;
}
.taskContainer{
  border: 1px solid blue;
  border-radius: 10px;
  padding: 15px;
}
.taskTitle{
  font-size: 16px;
  font-weight: 600;
  margin: 10px 0 5px 0;
}

.btnContainer{
@include  flexEndCenter;
  margin-top: 10px;
  button{
    border: none;
    font-weight: 600;
    border-radius: 5px;
    color: white;
    width: 100px;
    font-size: 15px;
  }
}

.removeBtn{
  width: 100%;
  margin-top: 20px;
  //border: 2px solid #ea6e40;
  border: none;
  font-weight: 600;
  border-radius: 5px;
  color: white;
  background-color: #ea6e40;
}

.removeBtn:hover{
  background-color: #c95f37;
}

.addBtn{
  width: 100%;
  margin-top: 20px;
  //border: 2px solid #ea6e40;
  border: none;
  font-weight: 600;
  border-radius: 5px;
  color: white;
  background-color: #509645;
}

.addBtn:hover{
  background-color: #396e30;
}
.cancelBtn{
  background-color: red;
  color: #000000;
  margin: 10px;
}

.saveBtn{
  background-color: blue;

}