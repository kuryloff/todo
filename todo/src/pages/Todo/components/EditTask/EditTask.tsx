import React, { useEffect, useState } from "react";
import { Task } from "../../../../interfaces/todoList.interface";
import styles from "./EditTask.module.scss";
import { Input } from "antd";
//@ts-ignore
import { v4 as uuidv4 } from "uuid";
interface Props {
  editMode: boolean;
  addMode: boolean;
  editItem: Task | undefined;
  setEditMode: (b: boolean) => void;
  setAddMode: (b: boolean) => void;
  handleSave: (b: Task) => void;
  handleCreateTask: (b: Task) => void;
}
const EditTask = ({
  editMode,
  editItem,
  setEditMode,
  handleSave,
  setAddMode,
  addMode,
  handleCreateTask,
}: Props) => {
  const [inputValues, setInputValue] = useState<Task>();
  useEffect(() => {
    setInputValue(editItem);
  }, [editItem]);

  const addSubtask = () => {
    const subtasks = inputValues?.subtasks;
    // @ts-ignore
    subtasks.push({
      id: uuidv4(),
      title: "",
      status: 0,
    });
    // @ts-ignore
    setInputValue({
      ...inputValues,
      // @ts-ignore
      subtasks,
    });
  };

  const removeSubtask = () => {
    const subtasks = inputValues?.subtasks;
    // @ts-ignore
    subtasks.pop();
    // @ts-ignore
    setInputValue({
      ...inputValues,
      // @ts-ignore
      subtasks,
    });
  };
  const editSubtask = (index: number, e: any) => {
    // @ts-ignore
    setInputValue({
      ...inputValues,
      // @ts-ignore
      subtasks: inputValues?.subtasks.map((task, i) => {
        if (index !== i) {
          return task;
        } else {
          task.title = e.target.value;
          return task;
        }
      }),
    });
  };

  // @ts-ignore
  return (
    <div className={styles.container} data-testid="EditTask">
      {inputValues && (editMode || addMode) && (
        <>
          <h1 className={styles.headerTitle}>
            {editMode ? "Edit" : "Add"} your Task
          </h1>
          <div className={styles.taskContainer}>
            <p className={styles.taskTitle}>Title</p>
            <Input
              allowClear
              placeholder="Task title"
              value={inputValues?.title}
              onChange={(e: any) =>
                setInputValue({
                  ...inputValues,
                  title: e.target.value,
                })
              }
            />
            <p className={styles.taskTitle}>Description</p>
            <Input
              allowClear
              placeholder="Task description"
              value={inputValues?.description}
              onChange={(e: any) =>
                setInputValue({
                  ...inputValues,
                  description: e.target.value,
                })
              }
            />
            {inputValues?.subtasks && inputValues?.subtasks?.length > 0 && (
              <>
                <p className={styles.taskTitle}>Subtasks</p>
                {inputValues?.subtasks.map((subtask, index) => (
                  <div key={subtask.id}>
                    <p className={styles.taskTitle}>Title</p>
                    <Input
                      allowClear
                      placeholder="Subtask title"
                      value={subtask?.title}
                      onChange={(e: any) => editSubtask(index, e)}
                    />
                  </div>
                ))}
              </>
            )}
            <button className={styles.addBtn} onClick={addSubtask}>
              Add Subtask
            </button>
            <button className={styles.removeBtn} onClick={removeSubtask}>
              Remove Subtask
            </button>
            <div className={styles.btnContainer}>
              <button
                className={styles.cancelBtn}
                onClick={() =>
                  editMode ? setEditMode(false) : setAddMode(false)
                }
              >
                Cancel
              </button>
              <button
                className={styles.saveBtn}
                onClick={() =>
                  editMode
                    ? handleSave(inputValues)
                    : handleCreateTask(inputValues)
                }
              >
                Save
              </button>
            </div>
          </div>
        </>
      )}
    </div>
  );
};
export default EditTask;
