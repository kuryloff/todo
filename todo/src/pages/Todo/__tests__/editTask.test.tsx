import { render, screen, cleanup } from "@testing-library/react";
import React from "react";
import EditTask from "../components/EditTask/EditTask";
import { defaultTasksList } from "../../../mockData/defaultTasksList";
import renderer from "react-test-renderer";

afterEach(() => {
  cleanup();
});
// @ts-ignore
test("should render EditTask Component with subtasks", () => {
  render(
    <EditTask
      editMode={true}
      addMode={false}
      editItem={defaultTasksList[1]}
      setEditMode={() => console.log("change")}
      setAddMode={() => console.log("change")}
      handleSave={() => console.log("change")}
      handleCreateTask={() => console.log("change")}
    />
  );
  const task = screen.getByTestId("EditTask");
  expect(task).toBeInTheDocument();
  expect(task).not.toContainHTML("small");
});

//todo need to update snapshot for running tests due to uuid
test("EditTask snapshot match", () => {
  const dom = renderer
    .create(
      <EditTask
        editMode={true}
        addMode={false}
        editItem={defaultTasksList[1]}
        setEditMode={() => console.log("change")}
        setAddMode={() => console.log("change")}
        handleSave={() => console.log("change")}
        handleCreateTask={() => console.log("change")}
      />
    )
    .toJSON();
  expect(dom).toMatchSnapshot();
});
