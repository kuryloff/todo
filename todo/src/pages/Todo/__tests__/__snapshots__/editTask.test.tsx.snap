// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditTask snapshot match 1`] = `
<div
  className="container"
  data-testid="EditTask"
/>
`;
